File: core/src/main/java/nl/sjtek/control/core/modules/Temperature.java
package nl.sjtek.control.core.modules;

import com.google.common.eventbus.Subscribe;
import nl.sjtek.control.core.events.Bus;
import nl.sjtek.control.core.settings.SettingsManager;
import nl.sjtek.control.data.ampq.events.TemperatureEvent;
import nl.sjtek.control.data.responses.Response;
import nl.sjtek.control.data.responses.TemperatureResponse;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.ResponseBody;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.*;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.*;

public class Temperature extends BaseModule {

    private static final int UPDATE_DELAY = 900000;
    private static final String WEATHER_URL_OUTSIDE = String.format("https://api.darksky.net/forecast/%s/51.5121298,5.4924242", SettingsManager.getInstance().getWeather().getApiKey());
    private static final String WEATHER_URL_INSIDE = "http://10.10.0.2/cgi-bin/temp";
    private static final String LOG_PATH = "/var/sjtekcontrol/log.csv";
    private static final int ID_INSIDE = 1;

    private final OkHttpClient httpClient = new OkHttpClient();

    private int errorsOutside = 0;
    private int tempInside = 0;
    private int tempOutside = -100;
    private float humidity = 0;
    private String description = "error";
    private String icon = "";

    public Temperature(String key) {
        super(key);
        Bus.regsiter(this);
        Timer updateTimer = new Timer();
        updateTimer.scheduleAtFixedRate(new UpdateTask(), 0, UPDATE_DELAY);
    }

    @Override
    public Response getResponse() {
        return new TemperatureResponse(tempOutside, tempInside, humidity, description, icon);
    }

    @Override
    public String getSummaryText() {
        return "The temperature inside is " + tempInside + " degrees, and outside " + tempOutside + " degrees.";
    }

    @Subscribe
    public void onTemperatureEvent(TemperatureEvent event) {
        System.out.println("ayyyyyy dikke aight " + event.toString());
        switch (event.getId()) {
            case ID_INSIDE:
                tempInside = (int) event.getTemperature();
                dataChanged(false);
                break;
        }
    }

    private void parseOutside(String response) {
        if (!response.isEmpty()) {
            try {
                // Powered by darksky.net :)
                JSONObject jsonObject = new JSONObject(response);
                JSONObject currently = jsonObject.getJSONObject("currently");
                float temp = currently.getLong("temperature");
                double humidity = currently.getDouble("humidity");
                String description = currently.getString("summary");

                temp = ((temp - 32) * 5) / 9;

                this.tempOutside = (int) temp;
                this.humidity = (float) humidity;
                this.description = description;
                this.icon = "";
                errorsOutside = 0;
                return;
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Temperature outside error: " + response);
        errorsOutside++;
        if (errorsOutside > 1) {
            tempOutside = -100;
        }
    }

    @Deprecated
    private void parseInside(String response) {
        if (!response.isEmpty()) {
            try {
                this.tempInside = (int) (float) Float.valueOf(response);
                return;
            } catch (NumberFormatException ignored) {
            }
        }
        tempInside = -100;
    }

    private String download(String stringUrl) {

        okhttp3.Response response = null;
        try {
            response = httpClient.newCall(new Request.Builder().url(stringUrl).build()).execute();
            System.out.println("" + response.code() + " - downloaded " + stringUrl);
            if (response.code() == 200) {
                ResponseBody body = response.body();
                return body.string();
            } else {
                return "";
            }
        } catch (IOException e) {
            return "";
        } finally {
            if (response != null) response.body().close();
        }
    }

    public String getLogData() {
        StringBuilder builder = new StringBuilder();
        builder.append("[");
        String rowTemplate = "[\"%s\", %s, %s]";
        try (BufferedReader reader = new BufferedReader(new FileReader(LOG_PATH))) {
            String line;
            String prefix = "";
            while ((line = reader.readLine()) != null) {
                String row[] = line.split(";");
                builder.append(prefix);
                prefix = ",";
                builder.append(String.format(rowTemplate, row[0], row[1].replace(',', '.'), row[2].replace(',', '.')));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        builder.append("]");
        return builder.toString();
    }

    private class UpdateTask extends TimerTask {

        @Override
        public void run() {
            // parseInside(download(WEATHER_URL_INSIDE));
            parseOutside(download(WEATHER_URL_OUTSIDE));

            try (BufferedWriter writer = new BufferedWriter(new FileWriter(LOG_PATH, true))) {
                TimeZone tz = TimeZone.getTimeZone("UTC");
                DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'"); // Quoted "Z" to indicate UTC, no timezone offset
                df.setTimeZone(tz);
                String nowAsISO = df.format(new Date());
                String data = String.format(Locale.GERMAN, "%s;%d;%d;\n",
                        nowAsISO, tempInside, tempOutside);
                writer.write(data);
            } catch (IOException e) {
                e.printStackTrace();
            }
            dataChanged();
        }
    }
}

File: core/src/main/kotlin/nl/sjtek/control/core/modules/Temperature.kt
package nl.sjtek.control.core.modules

import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import net.engio.mbassy.listener.Handler
import nl.sjtek.control.core.events.Bus
import nl.sjtek.control.core.events.TemperatureEvent
import nl.sjtek.control.core.net.HttpClient
import nl.sjtek.control.core.response.ResponseCache
import nl.sjtek.control.core.settings.SettingsManager
import nl.sjtek.control.data.response.Response
import nl.sjtek.control.data.response.Temperature
import okhttp3.Request
import org.slf4j.LoggerFactory
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

class Temperature(key: String) : Module(key) {
    private val logger = LoggerFactory.getLogger(javaClass)
    private val executor = ScheduledThreadPoolExecutor(1)
    override val response: Response
        get() = Temperature(key, insideTemp, insideHumidity, outsideTemp, outsideHumidity)

    private var insideTemp: Float = -100f
    private var insideHumidity: Float = -100f
    private var insideUpdate: Long = System.currentTimeMillis()
    private var outsideTemp: Float = -100f
    private var outsideHumidity: Float = -100f
    private var outsideUpdate: Long = System.currentTimeMillis()

    init {
        executor.scheduleAtFixedRate(this::updateInside, 0, 15, TimeUnit.MINUTES)
        executor.scheduleAtFixedRate(this::updateOutside, 0, 15, TimeUnit.MINUTES)
        Bus.subscribe(this)
    }

    override fun initSpark() {

    }

    @Handler
    fun onTemperatureUpdate(event: TemperatureEvent) {
        if (event.id == SENSOR_ID) {
            insideTemp = event.temperature
            insideHumidity = event.humidity
            insideUpdate = System.currentTimeMillis()
        }
        ResponseCache.post(this, true)
    }

    private fun updateInside() {
        if (System.currentTimeMillis() - insideUpdate > UPDATE_TIMEOUT) {
            insideTemp = -100f
            insideHumidity = -100f
        }
        ResponseCache.post(this, true)
    }

    private fun updateOutside() {
        logger.info("Updating temperature outside")
        val request = Request.Builder()
                .url(URL_OUTSIDE)
                .build()
        HttpClient.client.newCall(request).execute().use { response ->
            val jsonString = response.body()?.string() ?: ""
            try {
                val json = JsonParser().parse(jsonString).asJsonObject
                val currently = json.getAsJsonObject("currently")
                val temp = currently.get("temperature").asFloat
                val humidity = currently.get("humidity").asFloat

                outsideTemp = ((temp - 32) * 5 / 9)
                outsideHumidity = humidity
                outsideUpdate = System.currentTimeMillis()
            } catch (e: JsonParseException) {
                logger.error("Temperatue parsing error", e)
            }
        }

        if (System.currentTimeMillis() - outsideUpdate > UPDATE_TIMEOUT) {
            outsideTemp = -100f
            outsideHumidity = -100f
        }

        ResponseCache.post(this, true)
    }

    companion object {
        val SENSOR_ID = SettingsManager.settings.temperature.insideSensorId
        val URL_OUTSIDE = SettingsManager.settings.temperature.urlOutside.format(SettingsManager.settings.temperature.apiKey)
        const val UPDATE_TIMEOUT = 3000000 // 50 minutes
    }
}